Update TCK tests for OTel metrics types and units

Azquelt opened the pull request:
For #639, matching the spec changes proposed in #640

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/util/InMemoryMetricReader.java:
```diff
@@ -71,61 +87,139 @@ public CompletableResultCode shutdown() {
+    private static Predicate<PointData> hasAllAttributes(Attributes attributes) {
+        return pointData -> {
+            for (Entry<AttributeKey<?>, Object> e : attributes.asMap().entrySet()) {
+                if (!pointData.getAttributes().get(e.getKey()).equals(e.getValue())) {
```
This will NPE if `e.getKey()` isn't in the attributes.

Azquelt replied:
Good point, thank you

Azquelt replied:
Pushed a commit to fix this.

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/util/TelemetryMetricID.java:
```diff
@@ -30,59 +30,48 @@ public class TelemetryMetricID {
-        hash = hash * 31 + name.hashCode();
-
-        return hash;
+        return Objects.hash(attributes, name, type);
```
I wasn't aware of this bit of syntactic sweetness. Thanks for making me aware of it :)

Azquelt replied:
I'll be honest, Eclipse's "Generate hashCode and equals" that made me aware of it.

Azquelt replied:
I should also note that this is only valid because implementations of `Attributes` are required to implement `equals` and `hashCode` correctly, otherwise the code that you had which was doing its own hashing of `attributes` would have been necessary.

Often interfaces assert this by redeclaring `equals` and `hashCode` (e.g. `java.util.Map` does this), but in the case of `Attributes`, it's only stated [in the class description](https://www.javadoc.io/doc/io.opentelemetry/opentelemetry-api/1.39.0/io/opentelemetry/api/common/Attributes.html).

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/util/TelemetryHistogramMetric.java:
```diff
@@ -32,31 +42,126 @@
+
+        // Check we have the right number of results in the histogram:
+        assertEquals(pointData.getCount(), expectedResultsMillis.length,
+                "Wrong number of results in histogram getCounts for " + metricId.name);
```
Please can we have metricId.toString() for these three asserts.

Azquelt replied:
Done

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/util/TelemetryHistogramMetric.java:
```diff
@@ -32,31 +42,126 @@
+     * @param expectedResultsMillis
+     *            the list of values which the histogram is expected to have recorded
+     */
+    public void assertBucketCounts(long... expectedResultsMillis) {
```
This is a really nice method!

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/BulkheadTelemetryTest.java:
```diff
@@ -206,24 +203,24 @@ public void bulkheadMetricHistogramTest() throws InterruptedException, Execution
+                .orElse(0L);
 
-        assertThat("mean", Math.round(time / count), approxMillis(1000)); // histogram
+        assertThat("mean", time / count, approxMillisFromSeconds(1000)); // histogram
```
This could cause a divide by zero exception if `ft.bulkhead.runningDuration` isn't exported. Not the biggest issue since that's a fail, and it didn't cause problems in the old code, but since we have an Optional anyway we could assert isPresent for a nicer error message.

Azquelt replied:
Good point, we should stop the test if there's no data for the metric.

However, I think we still need to assert `count > 0` before computing the mean, since the metric could be present but still have no data in it.

Azquelt replied:
Hmm, in fact we're already asserting `count == 2` above but just getting it in a different way... let's tidy this up.

Azquelt replied:
Pushed a commit to tidy up this method.

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/BulkheadTelemetryTest.java:
```diff
@@ -206,24 +203,24 @@ public void bulkheadMetricHistogramTest() throws InterruptedException, Execution
         // Should have 4 results, ~0ms * 2 and ~1000ms * 2
-        executionTimesCount = m.getBulkheadRunningDuration().getHistogramCount().get();
-        assertThat("histogram count", executionTimesCount, is(4L));
+        m.getBulkheadRunningDuration().assertBucketCounts(1000, 1000, 0, 0);
```
I wonder if we should go back to the old metric test and also make that use buckets? Not urgent of course.

Azquelt replied:
That would be a good idea in the future, but at the moment the TCK compiles against MP Metrics 4.0 which doesn't have explicit buckets, it just has percentiles (which we can't reliably test, see message on 28d305).

If we set a minimum version of MP Metrics in a future release, we could add an assertion like this to `BulkheadMetricTest`.

benjamin-confino commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/telemetryMetrics/util/TelemetryHistogramMetric.java:
```diff
@@ -32,31 +42,126 @@
+
+    public void assertBoundaries() {
+        HistogramPointData pointData =
+                getHistogramPoint().orElseThrow(() -> new AssertionError("No data point for " + metricId.name));
```
Can this output metricId.toString() please.

Azquelt replied:
Done

Azquelt commented:
- Squashed review changes
- Reverted the dependency on MP Telemetry to RC2, so that the build checks will pass and we can release a new FT RC
- rebased on `main`
